src/console/CliManager.cpp

```diff
@@ -240,5 +243,313 @@ void CliManager::loadHistory() {
     ::fclose(file);
 }
 
+
+struct StringCaseEqual {
+    bool operator()(const std::string &lhs, const std::string &rhs) const {
+        return ::strcasecmp(lhs.c_str(), rhs.c_str()) == 0;
+    }
+};
+
+
+struct StringCaseHash {
+    size_t operator()(const std::string &lhs) const {
+        std::string upper;
+        upper.resize(lhs.size());
+        auto toupper = [] (auto c) { return ::toupper(c); };
+        std::transform(lhs.begin(), lhs.end(), upper.begin(), toupper);
+        return std::hash<std::string>()(upper);
+    }
+};
+
+// Primary keywords, like `GO' `CREATE', etc.
+static std::vector<std::string> primaryKeywords;
+
+// Keywords along with their sub-keywords, like `SHOW': `TAGS', `SPACES'
+static std::unordered_map<std::string, std::vector<std::string>,
+                          StringCaseHash, StringCaseEqual> subKeywords;
+// Typenames, like `int', `double', `string', etc.
+static std::vector<std::string> typeNames;
+
+
+// To fill the containers above from a json file.
+static Status loadCompletions();
+static Status parseKeywordsFromJson(const folly::dynamic &json);
+
+// To retrieve matches from within the `primaryKeywords'
+static std::vector<std::string>
+matchFromPrimaryKeywords(const std::string &text);
+
+// To retrieve matches from within the `subKeywords'
+static std::vector<std::string> matchFromSubKeywords(const std::string &text,
+                                                     const std::string &primaryKeyword);
+
+// Given a collection of keywords, retrieve matches that prefixed with `text'
+static std::vector<std::string> matchFromKeywords(const std::string &text,
+                                                  const std::vector<std::string> &keywords);
+
+// To tell if the current `text' is at the start position of a statement.
+// If so, we should do completion with primary keywords.
+// Otherwise, the primary keyword of the current statement
+// will be set, thus we will do completion with its sub keywords.
+static bool isStartOfStatement(std::string &primaryKeyword);
+
+// Given the prefix and a collection of keywords, retrieve the longest common prefix
+// e.g. given `u' as the prefix and [USE, USER, USERS] as the collection, will return `USE'
+static auto longestCommonPrefix(std::string prefix,
+                                const std::vector<std::string>& words);
+
+// Callback by realine if an auto completion is triggered
+static char** completer(const char *text, int start, int end);
```

Comment:
why add `start` and `end` ?

Reply:
This is a callback function, so they are required while we don't use them.